network-mux/src/Network/Mux.hs

```diff
@@ -228,27 +241,82 @@ muxStart tracer (MuxApplication ptcls) bearer = do
-        throwM e
-
-      -- These would always be internal errors, so propagate.
-      MuxerException   e -> throwM e
-      DemuxerException e -> throwM e
-    --TODO: decide if we should have exception wrappers here to identify
-    -- the source of the failure, e.g. specific mini-protocol. If we're
-    -- propagating exceptions, we don't need to log them.
+monitor tracer jobpool
+        (MiniProtocolDispatch _ dispatchtbl)
+        (MiniProtocolResponders respondertbl) =
+    go (Set.fromList . range . bounds $ respondertbl)
+  where
+    -- To do this second job it needs to keep track of which responder protocol
+    -- threads are running.
+    go :: Set MiniProtocolIx
+       -> m ()
+    go !idleResponders = do
+      result <- atomically $
+            -- wait for a mini-protocol thread to terminate
+            (Left  <$> JobPool.collect jobpool)
+            -- or wait for data to arrive on the channels that do not yet have
+            -- responder threads running
+        <|> (Right <$> foldr (<|>) retry
+                         [ checkNonEmptyBuf dispatchInfo >> return ix
+                         | ix <- Set.elems idleResponders
+                         , let dispatchInfo = dispatchtbl ! (ix, ModeResponder)
+                         ])
+
+      case result of
+        -- For now we do not restart protocols, when any stop we terminate
+        -- and the whole bundle will get cleaned up.
+        Left (MiniProtocolShutdown pnum _pix pmode) -> do
+          traceWith tracer (MuxTraceState Dead)
+          traceWith tracer (MuxTraceCleanExit pnum pmode)
+
+        Left (MiniProtocolException pnum _pix pmode e) -> do
+          traceWith tracer (MuxTraceState Dead)
+          traceWith tracer (MuxTraceExceptionExit pnum pmode e)
+          throwM e
+
+        -- These would always be internal errors, so propagate.
+        --TODO: decide if we should have exception wrappers here to identify
+        -- the source of the failure, e.g. specific mini-protocol. If we're
+        -- propagating exceptions, we don't need to log them.
+        Left (MuxerException   e) -> do
+          traceWith tracer (MuxTraceState Dead)
+          throwM e
+        Left (DemuxerException e) -> do
+          traceWith tracer (MuxTraceState Dead)
+          throwM e
+
+        -- Data has arrived on a channel for a mini-protocol that do not yet
+        -- have a responder thread running. So we start it now.
+        Right ix | Just job <- respondertbl ! ix -> do
+         JobPool.forkJob jobpool job
+         go (Set.delete ix idleResponders)
+
+        Right _ix ->
+          --TODO: it would be cleaner to do this check in the demuxer.
+          throwM $ MuxError MuxInitiatorOnly
+                      ("Received data on a responder channel but this mux "
+                       ++ "instance is initiator only")
+                   callStack
+
+    checkNonEmptyBuf (MiniProtocolDispatchInfo q _) = do
```

Comment:
A type signature for this would be nice.

Reply:
done